### PR TITLE
[4.1] Inject the event dispatcher into the MVC layer

### PR DESCRIPTION
**Pull request by laoneo**
Pull Request for discussion https://github.com/joomla/joomla-cms/pull/35396#discussion_r702235580.

### Summary of Changes
Injects a dispatcher into the MVC layer, so we can trigger the events trough it. It removes the need to use the static call on the Factory. A new helper function `dispatchEvent` is added to the `BaseDatabaseModel` and `AbstractView` which does a fallback to the global dispatcher when none is available. This is is handy as most of the events are sent out in models or views and it guarantees backwards compatibility when an extension uses the model or view outside of the MVC context.

**Before you did**

`Factory::getApplication()->triggerEvent('myEvent', $event);`

or 

`Factory::getApplication()->getDispatcher()->dispatch('myEvent', $event);`

**Now you can do**

`$this->getDispatcher()->dispatch('myEvent', $event);`

or

`$this->dispatchEvent($event);`

### Testing Instructions
- Save an article.
- Open an article on the front.

### Actual result BEFORE applying this Pull Request
Works.

### Expected result AFTER applying this Pull Request
Works.

### CODE REVIEW

**Comment by wilsonge** [comment]
I think this makes sense for controller and model - less so for the view.

**Comment by laoneo** [comment]
As you can see, every details view, like the articles view, is triggering events. There are more the controllers which do not fir events.

**Comment by RickR2H** [comment]
I have tested this item :white_check_mark: successfully on 2a4c5c0bd3f8919d445373eb2fea7ba4dafe5337<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35899">issues.joomla.org/tracker/joomla-cms/35899</a>.</sub>

**Comment by joomdonation** [comment]
I have tested this item :white_check_mark: successfully on 2a4c5c0bd3f8919d445373eb2fea7ba4dafe5337<br /><br />Applied patch, save article, access to article, perform some batch actions (as changed in this PR) and it is still working. So the PR is working as described.

(I don't understand our event system good enough to confirm this is right or wrong, I just follow the testing instructions to test)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35899">issues.joomla.org/tracker/joomla-cms/35899</a>.</sub>

**Comment by alikon** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35899">issues.joomla.org/tracker/joomla-cms/35899</a>.</sub>

**Comment by laoneo** [comment]
@bembelimen friendly reminder

**Comment by laoneo** [comment]
@bembelimen friendly reminder

File: components/com_content/src/View/Article/HtmlView.php [hunk 1]
```diff
@@ -237,7 +238,7 @@ public function display($tpl = null)
 
 		// Process the content plugins.
 		PluginHelper::importPlugin('content');
-		Factory::getApplication()->triggerEvent('onContentPrepare', array('com_content.article', &$item, &$item->params, $offset));
+		$this->dispatchEvent(new Event('onContentPrepare', array('com_content.article', &$item, &$item->params, $offset)));
```
Comment (bembelimen) [inline components/com_content/src/View/Article/HtmlView.php:241]: Why not migration all the events which are in the base model, view, controller?

Comment (laoneo) [reply components/com_content/src/View/Article/HtmlView.php:241]: The other ones are legacy which do use the result argument in the base event and work with the return value of the event. We can't migrate them in a BC way.

**Comment by bembelimen** [comment]
Could you please update your branch with the latest 4.1-dev, then I can merge it (I think you didn't check the checkbox that maintainer can do it).

**Comment by laoneo** [comment]
DOne

**Comment by bembelimen** [comment]
Thx

**Comment by PhilETaylor** [comment]
Somehow everyone missed that there is a completely undefined class in this PR... https://github.com/joomla/joomla-cms/pull/36200

**Pull request merged**